core/client/fs/src/main/java/alluxio/client/file/cache/CacheManagerWithShadowCache.java

```diff
@@ -0,0 +1,234 @@
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import alluxio.client.quota.CacheQuota;
+import alluxio.client.quota.CacheScope;
+import alluxio.metrics.MetricKey;
+import alluxio.metrics.MetricsSystem;
+
+import com.codahale.metrics.Counter;
+import com.google.common.hash.BloomFilter;
+import com.google.common.hash.Funnel;
+import com.google.common.hash.PrimitiveSink;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicIntegerArray;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicLongArray;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+/**
+ * A wrapper class of CacheManager with shadow cache.
+ */
+public class CacheManagerWithShadowCache implements CacheManager {
+  private static final int NUM_BLOOM_FILTER = 4;
+  private static final int NUM_BLOOM_FILTER_EXPECT_INSERTION = 20_000_000;
+  private static final double NUM_BLOOM_FILTER_FPP = .00819;
+  private final CacheManager mCacheManager;
+  private final AtomicReferenceArray<BloomFilter<PageId>> mBFs =
+      new AtomicReferenceArray<BloomFilter<PageId>>(new BloomFilter[NUM_BLOOM_FILTER]);
+  private final AtomicIntegerArray mObjInserted = new AtomicIntegerArray(new int[NUM_BLOOM_FILTER]);
+  private final AtomicLongArray mByteInserted = new AtomicLongArray(new long[NUM_BLOOM_FILTER]);
+  ScheduledExecutorService mScheduler = Executors.newScheduledThreadPool(0);
+  // working set size window, in unit of second
+  private int mWindow = 20;
+  private int mCurrentFilterIndex = 0;
+  private int mShadowCachePages = 0;
+  private long mShadowCacheBytes = 0;
+  private AtomicLong mShadowCachePageRead = new AtomicLong(0);
+  private AtomicLong mShadowCachePageHit = new AtomicLong(0);
+  private AtomicLong mShadowCacheByteRead = new AtomicLong(0);
+  private AtomicLong mShadowCacheByteHit = new AtomicLong(0);
+
+  /**
+   * @param cacheManager the real cache manager
+   * @param window the past window in which is the working set
+   */
+  public CacheManagerWithShadowCache(CacheManager cacheManager, int window) {
+    mWindow = window;
+    mCacheManager = cacheManager;
+    for (int i = 0; i < mBFs.length(); ++i) {
+      mBFs.set(i, BloomFilter.create(PageIdFunnel.FUNNEL, NUM_BLOOM_FILTER_EXPECT_INSERTION,
+          NUM_BLOOM_FILTER_FPP));
+    }
+    mScheduler.scheduleAtFixedRate(this::switchBloomFilter, 0, mWindow / 4, SECONDS);
+    // update working set size every sec
+    mScheduler.scheduleAtFixedRate(this::updateWorkingSetSize, 0, 1, SECONDS);
+  }
+
+  /**
+   * Stop to switch bloom filters and update working set size.
+   */
+  public void stopUpdate() {
+    mScheduler.shutdown();
+  }
+
+  @Override
+  public boolean put(PageId pageId, byte[] page, CacheScope cacheScope, CacheQuota cacheQuota) {
+    if (!mBFs.get(mCurrentFilterIndex).mightContain(pageId)) {
+      mBFs.get(mCurrentFilterIndex).put(pageId);
```

Comment:
Is this thread-safe? What if `mCurrentFilterIndex` gets updated between the lines (e.g. switchBloomFilter happens)? I would suggest
```suggestion
BloomFilter  bf = mBFs.get(mCurrentFilterIndex)
if (!bf.mightContain(pageId)) {
      bf.put(pageId);
```